src/iOS/Xamarin.iOS/Samples/Augmented Reality/ViewHiddenInfrastructureAR/PipePlacer.cs

```diff
@@ -0,0 +1,250 @@
+            NavigationController.PopViewController(true);
+            NavigationController.PushViewController(new PipeViewerAR() { _pipeGraphics = _pipesOverlay.Graphics.Select(x => new Graphic(x.Geometry)) }, true);
+        }
+
+        private void RedoButton_Clicked(object sender, EventArgs e)
+        {
+            // Redo if possible.
+            if (_sketchEditor.RedoCommand.CanExecute(null)) _sketchEditor.RedoCommand.Execute(null);
+        }
+
+        private void UndoButton_Clicked(object sender, EventArgs e)
+        {
+            // Undo if possible.
+            if (_sketchEditor.UndoCommand.CanExecute(null)) _sketchEditor.UndoCommand.Execute(null);
+        }
+
+        private async void AddSketch(object sender, EventArgs e)
+        {
+            // Enable the editing buttons.
+            _doneButton.Enabled = _undoButton.Enabled = _redoButton.Enabled = true;
+
+            // Prevent the user from adding pipes concurrently.
+            _addButton.Enabled = false;
+
+            // Get the geometry of the user drawn line.
+            Geometry geometry = await _sketchEditor.StartAsync(SketchCreationMode.Polyline);
+            _addButton.Enabled = true;
+
+            // Verify that the user has drawn a polyline.
+            if (!(geometry is Polyline))
+            {
+                return;
+            }
+
+            // Get the first point of the pipe.
+            MapPoint firstPoint = ((Polyline)geometry).Parts[0].StartPoint;
+            try
+            {
+                // Get the users selected elevation offset.
+                double elevationOffset = _elevationSlider.Value;
+
+                // Get the elevation of the geometry from the first point of the pipe.
+                double elevation = await _elevationSurface.GetElevationAsync(firstPoint);
+
+                // Create a polyline for the pipe at the selected altitude.
+                Polyline elevatedLine = GeometryEngine.SetZ(geometry, elevation + elevationOffset) as Polyline;
+
+                // Create a graphic for the pipe.
+                Graphic linegraphic = new Graphic(elevatedLine);
+                _pipesOverlay.Graphics.Add(linegraphic);
+
+                // Display a message with the pipes offset from the surface.
+                if (elevationOffset < 0)
+                {
+                    _helpLabel.Text = string.Format("Pipe added {0:0.0}m below surface", elevationOffset * -1);
+                }
+                else if (elevationOffset == 0)
+                {
+                    _helpLabel.Text = "Pipe added at ground level";
+                }
+                else
+                {
+                    _helpLabel.Text = string.Format("Pipe added {0:0.0}m above the surface", elevationOffset);
+                }
+
+                // Enable the view button once a pipe has been added to the graphics overlay.
+                _viewButton.Enabled = true;
+            }
+            catch (Exception)
+            {
```

Comment:
Maybe log this exception?

Reply:
I added a write to the console. Exception will only occur if something happens in GetElevationAsync. Add sketch can't be called until the elevation surface is already loaded.